### PR TITLE
chore(#8862): bulk script: no password changes of admin/medic users, normalize, allow passwords to be specified 

### PR DESCRIPTION
**Pull request by mrjones-plip**
# Description

This PR updates the existing `bulk-password-update-export.js` [script](https://github.com/medic/cht-core/blob/master/scripts/bulk-password-update-export.js)  to 
* not allow changing passwords for users `medic` or `admin`.  While `4.6.0`+ [defend against](https://github.com/medic/cht-core/issues/8096) this already, earlier 4.x versions do not.
* allows you to specify CSV input of users so they can have their passwords specified instead of randomly added
* adds the same warning to both scripts "ARE YOU SURE YOU WANT TO CONTINUE!" 
* removes `bulk-password-functions.js` 

per #8862

###  `bulk-password-update-export.js` 

Now supports the `--use_passes` flag.  If you pass it `false` it will generate random passwords via `generatePassword()`. If you pass it `true` it will read the CSV.  Passwords that are insifiently complex will get refused with a `400` response and `The password is too easy to guess. Include a range of types of characters to increase the score.` message.

Here's the help text:

```
Change passwords of users in user-password-change.txt, show results on screen 

Usage:
      node bulk-password-update-export.js -h | --help
      node bulk-password-update-export.js --url https://localhost --user medic --password adminPass

Options:
    -h --help     Show this screen.
    --url         The url for the instance being changed
    --user        The admin user this operation is run as
    --password    The password for the admin user 
    --use_passes  The passwords are in user-password-change.txt in or not. 
                  "true" - yes they're there in USER, PASSWORD format
                  "false" - no, no password, just usernames, one per line. Generate a random password for me.
```

###  `bulk-password-functions.js` 

this script has been removed as it's been deemed too dangerous.  At a later date, we can add in an `--all` flag on  `bulk-password-update-export.js`  if needed 

# Code review checklist

- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [x] Tested: Unit and/or e2e where appropriate
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8860-defend-admin-pass-change/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8860-defend-admin-pass-change/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8860-defend-admin-pass-change/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: scripts/bulk-password-change.js [hunk 1]
```diff
@@ -44,24 +45,20 @@ const options = {
...
-    postOptions.body = {
-      'password': 'Secret_1'
-    };
-    postOptions.uri = `${options.uri}/${user.username}`;
-    try {
-      await rpn.post(postOptions);
-    } catch (e) {
-      console.log('An error while updating the password - ', e.message);
-    }
+    await changeUserPass(user.username, 'Secret_1',  options);
```
Comment (garethbowen) [inline scripts/bulk-password-change.js:61]: I don't understand the use case for this script. When would someone want to update all passwords to the same insecure one? I think we should deprecate this use case and not recommend it for anyone.

It's convenient to have a "reset all" script, but the passwords should be randomly generated (as in the other script).

Let's just delete this script to be safe. Then if this is a useful feature still, work to implement an `--all` flag in the other script so people have to type something extra to do the more destructive option.

Comment (mrjones-plip) [reply scripts/bulk-password-change.js:61]: There [was a use case](https://github.com/medic/cht-core/pull/7080#pullrequestreview-648719995) that was needed for a QA to test back when. but fair! I'll un-refactor the changes in [9b2d197](https://github.com/medic/cht-core/pull/8874/commits/9b2d1976b5faa8c7f6531d1fc2c15c66fe6d0107), delete the dangerous one and resubmit

Comment (mrjones-plip) [reply scripts/bulk-password-change.js]: fixed in 32adfe536!

also @billwambua had some changes he wanted to handle quoted strings which he added with a `String.prototype.splitCSV()` call.

File: scripts/bulk-password-update-export.js [hunk 2]
```diff
@@ -106,30 +127,49 @@ const loadUsers = async () => {
...
-  };
-  postOptions.uri = `${options.uri}/${user}`;
-  try {
-    await rpn.post(postOptions);
-    console.log('SUCCESS', user, newPass);
-  } catch (e) {
-    console.log('ERROR', user, e.message);
+
+// Thanks https://stackoverflow.com/a/40672956
+String.prototype.splitCSV = function() {
```
Comment (garethbowen) [inline scripts/bulk-password-update-export.js]: :warning: Don't update the prototype... ever.

https://docs.communityhealthtoolkit.org/contribute/code/style-guide/#do-not-extend-built-in-prototypes

It's just as easy to call a regular named function.

File: scripts/bulk-password-update-export.js [hunk 3]
```diff
@@ -106,30 +127,49 @@ const loadUsers = async () => {
...
+  let matches = this.match(/(\s*"[^"]+"\s*|\s*[^,]+|,)(?=,|$)/g);
+  for (let n = 0; n < matches.length; ++n) {
+    matches[n] = matches[n].trim();
+    if (matches[n] === ',') matches[n] = '';
+  }
+  return matches;
+}
```
Comment (garethbowen) [inline scripts/bulk-password-update-export.js:139]: CSVs are notoriously hard to parse so let's not write a bespoke parser when we have the whole of npm at our fingertips. In fact, the `load_messages` script already uses a csv parser so just use that one!

File: scripts/bulk-password-update-export.js [hunk 4]
```diff
@@ -67,6 +69,25 @@ Do you want to continue? [y/N]
...
+  const admins = ['admin', 'medic'];
+  const adminError = new Error('403 - Cannot change password for "medic" or "admin" users.');
+  const postOptions = {...options};
+  postOptions.body = {
+    'password': password
+  };
+  postOptions.uri = `${options.uri}/${user}`;
+  try {
+    if (admins.includes(user)) {
+      throw adminError;
```
Comment (garethbowen) [inline scripts/bulk-password-update-export.js:82]: Hmmm interesting. I think because you've created this in line 74 the stacktrace will show line 74 and not 82. I would just inline it here so it's created only if/when necessary (which feels like a tiny performance win too!), and also means you can include the specific bad username in the error message, eg:

```suggestion
      throw new Error(`403 - Cannot change password for administrator user "${user}"`);
```

File: scripts/bulk-password-update-export.js [hunk 5]
```diff
@@ -67,6 +69,25 @@ Do you want to continue? [y/N]
   }));
 };
 
+const changeUserPass = async (user, password, options) => {
+  const admins = ['admin', 'medic'];
```
Comment (garethbowen) [inline scripts/bulk-password-update-export.js:73]: I would pull this out as a top level constant. There's no need to init the same array over and over again!

**Review by garethbowen — CHANGES_REQUESTED** [review]

File: scripts/bulk-password-update-export.js (hunk 2 above)
Comment (garethbowen) [reply scripts/bulk-password-update-export.js]: Hah! I see sonar bet me to it!

File: scripts/bulk-password-update-export.js (hunk 3 above)
Comment (billwambua) [reply scripts/bulk-password-update-export.js:139]: [Papaparse](https://www.papaparse.com/docs) is good for this too.

File: scripts/bulk-password-update-export.js (hunk 4 above)
Comment (mrjones-plip) [reply scripts/bulk-password-update-export.js]: fixed in f0c5b2171 and c237a2f66

File: scripts/bulk-password-update-export.js (hunk 3 above)
Comment (mrjones-plip) [reply scripts/bulk-password-update-export.js]: @garethbowen  FTW in 63025e3!

**Review by garethbowen — APPROVED** [review]

File: scripts/bulk-password-update-export.js [hunk 6]
```diff
@@ -1,29 +1,33 @@
...
+const csvSync = require('csv-parse/sync');
 
 const argv = minimist(process.argv.slice(2), {
   alias: {
     'h': 'help'
   }
 });
 
-if (argv.h|| !argv.url || !argv.user || !argv.password) {
-  console.log(`Change passwords of users in user-password-change.txt to random value, show results on screen
```
Comment (billwambua) [inline scripts/bulk-password-update-export.js:16]: `user-password-change.txt` should be `user-password-change.csv`

File: scripts/bulk-password-update-export.js [hunk 7]
```diff
@@ -95,40 +116,42 @@ const generatePassword = async () => {
...
+  }
+  const users = await loadUsers(use_passes);
+  const input = await areYouSure(users.length, extraWarning);
   if (input.toLowerCase() !== 'y') {
     process.exit(0);
   }
 
   const postOptions = {...options};
   for (const user of users) {
+    if (!use_passes) {
```
Comment (billwambua) [inline scripts/bulk-password-update-export.js:152]: You should check whether `user` is empty.

Sometimes a csv might have an empty row in the middle of the sheet.

When that happens the script still does the call for password change only to get the following error back from CouchDB.
```
ERROR  404 - {"error":"not_found","reason":"Database does not exist."}
```

It doesn't cause a crash but printing to screen a message like `Error: CSV row is empty` would be more intuitive.

**Review by billwambua — CHANGES_REQUESTED** [review]

File: scripts/bulk-password-update-export.js (hunk 6 above)
Comment (mrjones-plip) [reply scripts/bulk-password-update-export.js:16]: good call! fixed in 0bc2eba85

File: scripts/bulk-password-update-export.js (hunk 7 above)
Comment (mrjones-plip) [reply scripts/bulk-password-update-export.js:155]: gotcha! fixed in 2c6cba7a0 . 

We now return a more clear `404`:

```
ERROR  404 - Username is blank - check CSV and run again.
```
